core/src/main/java/com/expedia/adaptivealerting/core/detector/PewmaOutlierDetector.java

```diff
@@ -0,0 +1,186 @@
+     * Creates a new PEWMA anomaly detector. Initial mean is given by initValue and initial standard deviation is 0.
+     *
+     * @param initialAlpha    Smoothing parameter.
+     * @param beta            Outlier weighting parameter.
+     * @param trainingLength  How many iterations to train for.
+     * @param weakThreshold   Weak outlier threshold, in sigmas.
+     * @param strongThreshold Strong outlier threshold, in sigmas.
+     * @param initValue       Initial observation, used to set the first mean estimate.
+     */
+    public PewmaOutlierDetector(
+            double initialAlpha,
+            double beta,
+            int trainingLength,
+            double weakThreshold,
+            double strongThreshold,
+            double initValue
+    ) {
+        isBetweenExclusive(initialAlpha, 0.0, 1.0, "initialAlpha must be in the range (0, 1)");
+
+        this.alpha0 = 1 - initialAlpha;  // To standardise with the EWMA implementation we use the complement value.
+        this.beta = beta;
+        this.trainingLength = trainingLength;
+        this.trainingCount = 1;
+        this.weakThreshold = weakThreshold;
+        this.strongThreshold = strongThreshold;
+        this.s1 = initValue;
+        this.s2 = pow(initValue, 2);
+        updateMeanAndStdDev();
+    }
+
+    private void updateMeanAndStdDev() {
+        this.mean = this.s1;
+        this.stdDev = sqrt(this.s2 - pow(this.s1, 2));
+    }
+    
+    @Override
+    public OutlierLevel evaluate(Instant instant, double value) {
+        
+        // TODO Currently we just ignore the instant. Would be helpful to have a way to detect missing data points in
+        // the stream, and to define strategies for dealing with this. [WLW]
+
+        final double dist = abs(value - mean);
+        final double largeThreshold = strongThreshold * stdDev;
+        final double smallThreshold = weakThreshold * stdDev;
+
+        OutlierLevel level = OutlierLevel.NORMAL;
+        if (dist > largeThreshold) {
+            level = OutlierLevel.STRONG;
+        } else if (dist > smallThreshold) {
+            level = OutlierLevel.WEAK;
+        }
+
+        updateEstimates(value);
+
+        return level;
+    }
+
+    private void updateEstimates(double value) {
+        double zt = 0;
+        if (this.stdDev != 0) {
+            zt = (value - this.mean)/this.stdDev;
+        }
+        double pt = (1/sqrt(2*Math.PI)) * exp(-(pow(zt, 2)/2));
+        double alpha = calculateAlpha(pt);
+
+        this.s1 = alpha*this.s1 + (1 - alpha)*value;
+        this.s2 = alpha*this.s2 + (1 - alpha)*(pow(value, 2));
+
+        updateMeanAndStdDev();
+    }
+
+    private double calculateAlpha(double pt) {
+        if (this.trainingCount < this.trainingLength) {
+            this.trainingCount++;
+            return 1 - 1.0/this.trainingCount;
+        }
+        return (1 - this.beta*pt)*this.alpha0;
+    }
+
+    public double getMean() {
+        return mean;
+    }
+
+    public double getStdDev() {
```

Comment:
Not sure if we want to standardise the getters. Could square it and return variance.

Reply:
I'm OK with aligning EWMA and PEWMA since they are pretty similar. I don't remember if I was returning stdev or variance in that case. It's probably too early to try to standardize this beyond these two detectors since this state is specific to these specific models. (For example with Aquila we do have a midpoint and dispersion, but they aren't necessarily estimates of mean and stdev.)

Whether we return stdev or variance should probably be driven by which internal representation is more directly useful for the calculations, given the online nature of the algo. In other words we wouldn't want to keep moving back and forth with each update. I think that's stdev here given how you are calculating `zt`.

Reply:
Might wait for a use-case that requires it, in PEWMA we keep stdDev in EWMA we keep variance.